qiskit_experiments/library/randomized_benchmarking/rb_experiment.py

```diff
@@ -206,284 +233,136 @@ def _sequences_to_circuits(
-        )
-        circ.barrier(qubits)
-        return circ, composed_cliff_num
-
-    def _add_inverse_to_circ(self, rb_circ, composed_num, qubits, clbits):
-        """Append the inverse of a circuit to the end of the circuit"""
-        inverse_cliff = StandardRB._clifford_utils.inverse_cliff(composed_num)
-        rb_circ.compose(inverse_cliff, inplace=True)
-        rb_circ.measure(qubits, clbits)
-        return rb_circ
-
-    # This method does a quick layout to avoid calling 'transpile()' which is
-    # very costly in performance
-    # We simply copy the circuit to a new circuit where we define the mapping
-    # of the qubit to the single physical qubit that was requested by the user
-    # This is a hack, and would be better if transpile() implemented it.
-    # Something similar is done in ParallelExperiment._combined_circuits
-    def _layout_for_rb(self):
-        transpiled = []
-        qargs_map = (
-            {0: self.physical_qubits[0]}
-            if self.num_qubits == 1
-            else {0: self.physical_qubits[0], 1: self.physical_qubits[1]}
-        )
-        for circ in self.circuits():
-            new_circ = QuantumCircuit(
-                *circ.qregs,
-                name=circ.name,
-                global_phase=circ.global_phase,
-                metadata=circ.metadata.copy(),
-            )
-            clbits = circ.num_clbits
-            if clbits:
-                creg = ClassicalRegister(clbits)
-                new_cargs = [Clbit(creg, i) for i in range(clbits)]
-                new_circ.add_register(creg)
-
-            for inst, qargs, cargs in circ.data:
-                mapped_cargs = [new_cargs[circ.find_bit(clbit).index] for clbit in cargs]
-                mapped_qargs = [circ.qubits[qargs_map[circ.find_bit(i).index]] for i in qargs]
-                new_circ.data.append((inst, mapped_qargs, mapped_cargs))
-                # Add the calibrations
-                for gate, cals in circ.calibrations.items():
-                    for key, sched in cals.items():
-                        new_circ.add_calibration(gate, qubits=key[0], schedule=sched, params=key[1])
-
-            transpiled.append(new_circ)
-        return transpiled
+        return op.adjoint()
 
     def _transpiled_circuits(self) -> List[QuantumCircuit]:
         """Return a list of experiment circuits, transpiled."""
-        if self.num_qubits in [1, 2]:
-            transpiled = self._layout_for_rb()
-        else:
+        has_custom_transpile_option = (
+            any(
+                opt not in {"basis_gates", "optimization_level"}
+                for opt in vars(self.transpile_options)
+            )
+            or self.transpile_options.get("optimization_level", 0) != 0
+        )
+        if self.num_qubits > 2 or has_custom_transpile_option:
             transpiled = super()._transpiled_circuits()
-        if self.analysis.options.get("gate_error_ratio", None) is None:
+        else:
+            transpiled = [
```

Comment:
Do you think worth having multi-threading here? Note that the standard Qiskit transpiler uses qiskit parallel and this is single thread.

Reply:
Yeah, I think it's worth investigating.

Reply:
I've examined parallel transpilation using `qiskit.tools.parallel.parallel_map` and found that it's much slower than serial one. I guess that is due to the overhead of pickling circuits. Let me keep the current code as is.

Examined code for parallel transpilation:
```
            circuits = self.circuits()
            if (
                    len(circuits) > 1
                    and os.getenv("QISKIT_IN_PARALLEL", "FALSE") == "FALSE"
                    and parallel.PARALLEL_DEFAULT
            ):
                # Custom transpilation in parallel
                transpiled = parallel.parallel_map(
                    _transpile_clifford_circuit,
                    circuits,
                    task_kwargs={"physical_qubits": self.physical_qubits},
                )
```

Reply:
This is good to know.

Reply:
Or maybe because I'm using Mac (I somewhere heard parallel does multi-processing so slow in Mac)

Reply:
On mac by default `parallel_map` won't actually run in parallel on Python >= 3.8 (https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/tools/parallel.py#L61-L83 defines the defaults) because in Python 3.8 they switched macOS `multiprocessing` to use `spawn` instead of `fork` by default because of some platform bugs. Spawn has a lot of extra overhead and it's also unreliable for user code because you need a `if __name__ == "__main__"` guard many times to distinguish a child process from the parent so we disable multiprocessing by default on macOS.

In general multiprocessing in python can be very tricky to work with and get a performance boost. Although for `transpile()` with linux (so using `fork()`) with involved compilation the performance is typically better even with the pickling and IPC overhead of separate processes (with some caveats https://github.com/Qiskit/qiskit-terra/issues/7741 )